### ISSUE TITLE
Enable thread sanitize vmfb on macOS

### ISSUE THREAD

**Issue by wangkuiyi** [iree-org/iree#12533]
### What happened?

When I was debugging a problem with fine-tuning of GPT-2 on macOS, iOS Simulator, and iPhone 12/13, I tried to thread-sanitize the iree-compile result. https://github.com/openxla/iree/issues/12529

Unfortunately, after enabling thread-sanitizing, the `iree-compile` command failed with `ld` errors: https://github.com/openxla/iree/issues/12529#issuecomment-1457475609

After some digging out https://github.com/openxla/iree/issues/12529#issuecomment-1457545830, Benoit and I realized that we need to add the following `ld` options 
```
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/clang/14.0.0/lib/darwin/libclang_rt.tsan_osx_dynamic.dylib \
  -rpath @executable_path \
  -rpath /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/clang/14.0.0/lib/darwin \
```
to https://github.com/openxla/iree/blob/1148f720be7e267f248e034b3cfb488633884980/compiler/src/iree/compiler/Dialect/HAL/Target/LLVM/internal/UnixLinkerTool.cpp#L82-L92

### What component(s) does this issue relate to?

Compiler

**Comment by wangkuiyi** [comment]
The following patch enabled the IREE compiler on my MacBook Pro generating TSAN-enabled vmfb:
```diff
       // Produce a Mach-O dylib file.
       flags.push_back("-dylib");
+      // For flat namespace see https://stackoverflow.com/a/49392862/724872.
       flags.push_back("-flat_namespace");
+      // Link to the C runtime library.
       flags.push_back(
           "-L /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/lib "
           "-lSystem");
+      flags.push_back(
+          "-L "
+          "/Applications/Xcode.app/Contents/Developer/Toolchains/"
+          "XcodeDefault.xctoolchain/usr/lib/clang/14.0.0/lib/darwin "
+          "-lclang_rt.osx");
```

The command-line is
```bash
iree-compile \
 ./runtime/src/iree/runtime/testdata/simple_mul.mlir \
 --iree-llvmcpu-link-embedded=false \
 --iree-llvmcpu-target-triple=arm64-apple-ios-simulator \
 --iree-input-type=mhlo \
 --iree-hal-target-backends=llvm-cpu \
 -o /tmp/mul.vmfb
```

However, this is very hacky and not something I can file a pull request.

1. `/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/lib` works because it happend that I downloaded and installed Apple Command Line Tools. Most macOS/iOS developers should not have done this, becaue the official solution is to install Xcode from App Store, which will give them `/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/lib` instead.  

   I have two Xcode of different versions and the Command Line Tools on my MacBook Pro, so I have the following three.
   1. `/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/`
   2. `/Applications/Xcode2.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/`
   3. `/Library/Developer/CommandLineTools/SDKs/`

   Apple users can switch among the above by running `xcode-select -s <path-to-one-of-the-above>`.  Different versions of macOS stores the selection in different places (https://stackoverflow.com/a/28533120/724872).
   
   **The official way to check the selection is to run the command `xcode-select -p`.**

1. `/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/clang/14.0.0/lib/darwin` is the one I chose on my MacBook Pro.

   **The official way to check the toolchain is to run `clang --version`.** For example,
   ```
   $ clang --version
   Apple clang version 14.0.0 (clang-1400.0.29.202)
   Target: arm64-apple-darwin22.3.0
   Thread model: posix
   InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
   ```
   
   Then, we will need to replace the suffix `/bin` by `/lib/clang`. This will give us a directory where there is a subdirectory named by the version of Clang that comes with the Xcode or Command Line Tools.

**Comment by benvanik** [comment]
hmm that makes cross compiling for iOS tricky - we don't want the clang reported version or what the host tools used to build the compiler are but the target of the vmfb. Letting the user choose this via flags or environment variables would be ideal. Another solution would be to shell exec `xcode-select -p` to try to get the directory if the user doesn't specify it manually. We've needed to fork UnixLInkerTool out into an AppleLinkerTool for awhile and such a detection path would make that really worth it!

Along with some other things we've seen this may also finally justify adding a flag like `--iree-llvmcpu-linker-flags=` that just appends to that flags list - then no need to modify code to set weird linker options/paths/etc. Using the system linker isn't on the default path for most users so being a little tricky is fine. Global flags also don't support multitargeting but if this is just used for debugging and not real deployment that's fine. We still of course want ASAN/TSAN/etc to work automatically though so solving that like above would still be good.

**Comment by wangkuiyi** [comment]
To complete the information -- there is an API in Objective-C and Swift that returns the chosen development tool; So, we can call
```objc
[NSWorkspace sharedWorkspace]fullPathForApplication:@"Xcode"];
```
However, I am afraid that calling Objective-C from our current C++ code might be more tricky than [`system`](https://man7.org/linux/man-pages/man3/system.3.html).

**Comment by benvanik** [comment]
yeah, that doesn't solve the cross compiling issue - we don't want to know where xcode building the compiler/runtime is but what the target is.
Neither does xcode-select really as we don't want users to have to xcode-select each time they iree-compile for a different target. 

I think the linker-flags are the first step (would unblock this and also make other experimentation easier) and then a dedicated flag for the toolchain path would solve the other - a user could put `xcodeselect -p` in their command line if they wanted to make it scriptable. Given that this is off the main path I think that's fine vs adding the weight of internal compiler selection.

**Comment by wangkuiyi** [comment]
I like the flag idea. Is the following command-line what we want?

```bash
iree-compile \
 ./runtime/src/iree/runtime/testdata/simple_mul.mlir \
 --iree-llvmcpu-link-embedded=false \
 --iree-llvmcpu-linker-flags="-L $(xcodeselect -p)../usr/lib/clang/...." \
 --iree-llvmcpu-target-triple=arm64-apple-ios-simulator \
 --iree-input-type=mhlo \
 --iree-hal-target-backends=llvm-cpu \
 -o /tmp/mul.vmfb
 ```

Am I supposed to add the flag `--iree-llvmcpu-linker-flags` at somewhere around [here](https://github.com/openxla/iree/blob/338b979201804353897f0a5c5f249b27fb62b7bc/compiler/src/iree/compiler/Dialect/HAL/Target/LLVMCPU/LLVMTargetOptions.cpp#L82-L95)?

**Comment by benvanik** [comment]
Yep! Down below we have linker related flags and it can go there: https://github.com/openxla/iree/blob/338b979201804353897f0a5c5f249b27fb62b7bc/compiler/src/iree/compiler/Dialect/HAL/Target/LLVMCPU/LLVMTargetOptions.cpp#L170

**Comment by wangkuiyi** [comment]
Why did we add https://github.com/openxla/iree/blob/5008b86e8795700daa03f814c22ffb192c76e576/compiler/src/iree/compiler/Dialect/HAL/Target/LLVMCPU/internal/UnixLinkerTool.cpp#L89

It gives me 
```
ld: warning: -flat_namespace is deprecated on free standing
```
